optimade/server/middleware.py

```diff
@@ -72,6 +75,130 @@ async def dispatch(self, request: Request, call_next):
+            return None
+
+        api_hint = f"/{api_hint[0]}"
+        if re.match(r"^/v[0-9]+(\.[0-9]+)?$", api_hint) is None:
+            warnings.warn(
+                FieldValueNotRecognized(
+                    detail=f"{api_hint[1:]!r} is not recognized as a valid `api_hint` value."
+                )
+            )
+            return None
+
+        if api_hint in BASE_URL_PREFIXES.values():
+            return api_hint
+
+        major_api_hint = int(re.findall(r"/v([0-9]+)", api_hint)[0])
+        major_implementation = int(BASE_URL_PREFIXES["major"][len("/v") :])
+
+        if major_api_hint > major_implementation:
+            # Let's not try to handle a request for a newer major version
+            raise VersionNotSupported(
+                detail=(
+                    f"The provided `api_hint` ({api_hint[1:]!r}) is not supported by this implementation. "
+                    f"Supported versions include: {', '.join(BASE_URL_PREFIXES.values())}"
+                )
+            )
+        if major_api_hint <= major_implementation:
+            # If less than:
+            # Use the current implementation in hope that it can still handle older requests
+            #
+            # If equal:
+            # Go to /v<MAJOR>, since this should point to the latest available
+            return BASE_URL_PREFIXES["major"]
+
+    @staticmethod
+    def is_versioned_base_url(url: str) -> bool:
+        """Determine whether a URL is for a versioned base URL
+
+        First, simply check whether a `/vMAJOR(.MINOR.PATCH)` part exists in the URL.
+        If not, return False.
+        Else, remove unversioned base URL from URL and check again.
+        Return bool of final result.
+        """
+        from optimade.server.routers.utils import get_base_url
+
+        if not re.findall(r"(/v[0-9]+(\.[0-9]+){0,2})", url):
+            return False
+
+        base_url = get_base_url(urllib.parse.urlparse(url))
+        return bool(re.findall(r"(/v[0-9]+(\.[0-9]+){0,2})", url[len(base_url) :]))
+
+    async def dispatch(self, request: Request, call_next):
+        parsed_query = urllib.parse.parse_qs(request.url.query, keep_blank_values=True)
+
+        if "api_hint" in parsed_query:
+            if self.is_versioned_base_url(str(request.url)):
+                warnings.warn(
+                    QueryParamNotUsed(
+                        detail=(
+                            "`api_hint` provided with value{:s} '{:s}' for a versioned base URL. "
+                            "In accordance with the specification, this will not be handled by "
+                            "the implementation.".format(
+                                "s" if len(parsed_query["api_hint"]) > 1 else "",
+                                "', '".join(parsed_query["api_hint"]),
+                            )
+                        )
+                    )
+                )
+            else:
+                from optimade.server.routers.utils import get_base_url
```

Comment:
Same as above. If there isn't a reason to lazy-load this import, can we just move it to the top?